File: androidApp/src/main/java/com/example/kmpsninotes/android/presentation/screen/login/LoginScreenViewModel.kt
package com.example.kmpsninotes.android.presentation.screen.login

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.domain.model.Login
import com.example.domain.model.User
import com.example.domain.repository.UserRepository
import com.example.kmpsninotes.android.until.Constants.AuthorizationMode
import com.example.kmpsninotes.android.until.Constants.RegistrationMode
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class LoginScreenViewModel @Inject constructor(
    private val userRepository: UserRepository
):ViewModel() {

    private val _uiState = MutableStateFlow(LoginUiState())
    val uiState:StateFlow<LoginUiState> = _uiState




    fun onEmailTextFieldChange(text:String){
        _uiState.value = uiState.value.copy(emailTextFieldValue = text)
    }

    fun onNameTextFieldChange(text:String){
        _uiState.value = uiState.value.copy(nameTextFieldValue = text)
    }
    fun onPasswordTextFieldChange(text:String){
        _uiState.value = uiState.value.copy(passwordTextFieldValue = text)
    }
    fun switchMode(){
        if(_uiState.value.mode == RegistrationMode){
            _uiState.value = uiState.value.copy(mode = AuthorizationMode, error = "")
        }else{
            _uiState.value = uiState.value.copy(mode = RegistrationMode, error = "")
        }
    }

    fun onRegSubmit(){
        viewModelScope.launch {
            val user = userRepository.registerUser(
                User(
                login = uiState.value.emailTextFieldValue,
                password = uiState.value.passwordTextFieldValue,
                name = uiState.value.nameTextFieldValue
              )
            )
            Log.d("efgfdwefgfd",user.data.toString())
            if(user.data != null){
                userRepository.putUserData(user.data!!)
                _uiState.value = uiState.value.copy(hasBeenDone = true)
            }else{
                Log.d("efgbgfwergfd",user.message.toString())
                _uiState.value = uiState.value.copy(error = user.message!!)
            }
        }
    }
    fun onAuthSubmit(){
        viewModelScope.launch {
            val session = userRepository.authUser(
                Login(
                login = uiState.value.emailTextFieldValue,
                password = uiState.value.passwordTextFieldValue
            )
            )
            if (session.data != null){
                val user = userRepository.observeUserData(
                    session = session.data!!,
                    email = uiState.value.emailTextFieldValue
                )
                 userRepository.putUserData(user.data!!)
                _uiState.value = uiState.value.copy(hasBeenDone = true)
            }else{
                _uiState.value = uiState.value.copy(error = session.message!!)
            }
        }
    }
}

File: iosApp/iosApp/login_screen/LoginScreenViewModel.swift
//
//  LoginScreenViewModel.swift
//  iosApp
//
//  Created by MacBook on 05.09.2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import shared


    
class LoginScreenViewModel:ObservableObject{
    @Published var login = ""
    @Published var name = ""
    @Published var password = ""
    
    @Published var isSignUp = false
    @Published var error:String? = nil
    
    @Published var hasBeenDone = false
    
    private var userRepository:UserRepository? = nil
    
    init(userRepository: UserRepository? = nil) {
        self.userRepository = userRepository
    }
    
    func loginUser(){
        
        if isSignUp{
            let user = User(
                login: login,
                password: password,
                session: "",
                name: name
            )
            userRepository?.registerUser(
                user: user,
                completionHandler: { res, err in
                    self.error = res?.message
                    if res!.data != nil{
                        self.userRepository?.putUserData(user: user, completionHandler: {_ in
                            self.hasBeenDone = true
                        })
                    }
                }
            )
        }else{
            userRepository?.authUser(
                login: Login(login: login, password: password),
                completionHandler: {res, err in
                    self.error = res?.message
                    
                    if res?.message != nil{
                        self.hasBeenDone = res!.message!.isEmpty
                    }else{
                        self.userRepository?.observeUserData(session: res?.data as! String, email: self.login, completionHandler: {user, err in
                            if user?.data != nil{
                                self.userRepository?.putUserData(user: user?.data ?? User(login: "",password: "",session: "",name: ""), completionHandler: {_ in
                                    self.hasBeenDone = true
                                    })
                                }
                            }
                        )
                    }
                }
            )
        }
       
    }
    
    func switchLoginMode(){
        isSignUp = !isSignUp
    }
    
}
    
